### DISCUSSION TITLE
Smart gem installation

### DISCUSSION THREAD

**Discussion by hurricup** [ruby/rubygems#5994] [Q&A] [answered]
From time to time we face a problem with gems installation when ruby or gem version is pretty old. This is caused by vague dependencies description and/or `gem` dumbness. 

If you try to install `rails` with `gem install rails` you may face an error that some dependency requires ruby XXX, last compatible version is YYY, please try to install with `gem install ZZZ -version YYY`

Probably this is not a big deal for occasional manual installation, but a really big pain for automatic installations.
It would be nice to be able to run `gem install` with some `--smart` or `--bundler` option, so bundler logic would be used to resolve dependencies, not just dumb constraints checking. 

Or to be able to run `bundle install-gem gem_name` for the same reason. Can't say which one is more appropriate here. 

Or may be there is a way to do this with cli or simple code?

Also, recently met unclear issue on the macos, after `sqlite3` 1.5.3 release. 
There are:
- sqlite3 1.5.3-ruby, with ruby >= 1.9.2 constraint
- sqlite3 1.5.3-x86_64-darwin with >= 2.6, < 3.2.DEV constraints

And on ruby 2.3 & 2.4 bundler attempts to install the last one and fails (dependency specified as `gem 'sqlite3'`. Probably this is caused by the old bundler version, checking this with updated `--system`

**Comment by deivid-rodriguez** [comment, marked as answer]
This is already the default behavior. The problem is that on old Rubies, the version of the `gem` CLI is too old and doesn't have this fix. If you run `gem update --system` first, `gem install` should resolve to the highest version that's also compatible with your current Ruby version.

**Reply by hurricup** [reply]
Wow, that's cool!!! What version this smart behavior was introduced in?

**Reply by simi** [reply]
https://github.com/rubygems/rubygems/blob/master/CHANGELOG.md#323--2020-12-22

**Reply by hurricup** [reply]
Thank you. 
Hm, this is a pretty old, feels like I had issues after it, but not 100% sure. Will give it a try.
